ako-gateway-api/objects/gateway_store.go

```diff
@@ -571,31 +638,103 @@ func (g *GWLister) DeleteRouteChildVSMappings(routeTypeNsName, childVS string) {
 	defer g.gwLock.Unlock()
 
 	// delete route to child vs mapping
-	_, childVSList := g.GetRouteToChildVS(routeTypeNsName)
-	childVSList = utils.Remove(childVSList, childVS)
-	g.routeToChildVS.AddOrUpdate(routeTypeNsName, childVSList)
+	if found, childVSList := g.routeToChildVS.Get(routeTypeNsName); found {
+		childVSListObj := childVSList.([]string)
+		childVSListObj = utils.Remove(childVSListObj, childVS)
+		if len(childVSListObj) == 0 {
+			g.routeToChildVS.Delete(routeTypeNsName)
+		} else {
+			g.routeToChildVS.AddOrUpdate(routeTypeNsName, childVSListObj)
+		}
+	}
 }
 
 //=====All route functions go here.
 
-func (g *GWLister) DeleteRouteGatewayMappings(routeTypeNsName string) {
+func (g *GWLister) DeleteRouteFromStore(routeTypeNsName string) {
 	g.gwLock.Lock()
 	defer g.gwLock.Unlock()
 
-	// delete route to gateway mapping
-	found, obj := g.routeToGateway.Get(routeTypeNsName)
-	if !found {
-		return
+	//delete route to gateway listener
+	g.routeToGatewayListener.Delete(routeTypeNsName)
+
+	//delete route to gateway
+	found, gatewayList := g.routeToGateway.Get(routeTypeNsName)
+	var gatewayListObj, svcListObj []string
+	if found {
+		gatewayListObj = gatewayList.([]string)
+	}
+	for _, gwNsName := range gatewayListObj {
+		if found, routeNsNameList := g.gatewayToRoute.Get(gwNsName); found {
+			routeNsNameListObj := routeNsNameList.([]string)
+			routeNsNameListObj = utils.Remove(routeNsNameListObj, routeTypeNsName)
+			if len(routeNsNameListObj) == 0 {
+				g.gatewayToRoute.Delete(gwNsName)
+			} else {
+				g.gatewayToRoute.AddOrUpdate(gwNsName, routeNsNameListObj)
+			}
+		}
 	}
-	gwNsNameList := obj.([]string)
 	g.routeToGateway.Delete(routeTypeNsName)
-	g.routeToGatewayListener.Delete(routeTypeNsName)
 
-	for _, gwNsName := range gwNsNameList {
-		if found, obj := g.gatewayToRoute.Get(gwNsName); found {
-			routeTypeNsNameList := obj.([]string)
-			routeTypeNsNameList = utils.Remove(routeTypeNsNameList, routeTypeNsName)
-			g.gatewayToRoute.AddOrUpdate(gwNsName, routeTypeNsNameList)
+	//delete route to service
+	found, svcList := g.routeToService.Get(routeTypeNsName)
+	if found {
+		svcListObj = svcList.([]string)
+	}
+	for _, svcNsName := range svcListObj {
+		if found, routeNsNameList := g.serviceToRoute.Get(svcNsName); found {
+			routeNsNameListObj := routeNsNameList.([]string)
+			routeNsNameListObj = utils.Remove(routeNsNameListObj, routeTypeNsName)
+			if len(routeNsNameListObj) == 0 {
+				g.serviceToRoute.Delete(svcNsName)
+			} else {
+				g.serviceToRoute.AddOrUpdate(svcNsName, routeNsNameListObj)
+			}
+		}
+	}
+	g.routeToService.Delete(routeTypeNsName)
+
+	//update gateway to service mappings after route deletion
```

Comment:
What is the reason of keeping Service to Gateway mapping and vice versa?

Reply:
when a service gets updated, we need to know what are the affected graph objects associated with gateways